chains/ethereum/connection.go

```diff
@@ -0,0 +1,174 @@
+// Copyright 2020 ChainSafe Systems
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package ethereum
+
+import (
+	"context"
+	"fmt"
+	"math/big"
+	"sync"
+
+	"github.com/ChainSafe/ChainBridge/crypto/secp256k1"
+	"github.com/ChainSafe/log15"
+
+	eth "github.com/ethereum/go-ethereum"
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	ethcommon "github.com/ethereum/go-ethereum/common"
+	ethtypes "github.com/ethereum/go-ethereum/core/types"
+	ethcrypto "github.com/ethereum/go-ethereum/crypto"
+	"github.com/ethereum/go-ethereum/ethclient"
+	"github.com/ethereum/go-ethereum/rpc"
+)
+
+// Nonce is a struct that wraps the Nonce with a mutex lock
+// this struct was implemented to prevent race conditions where
+// two transactions try to transact at the same time and recieve
+// the same nonce, causing one to be rejected.
+type Nonce struct {
+	nonce uint64
+	lock  *sync.Mutex
+}
+
+type Connection struct {
+	cfg       Config
+	ctx       context.Context
+	conn      *ethclient.Client
+	signer    ethtypes.Signer
+	kp        *secp256k1.Keypair
+	nonceLock sync.Mutex
+	log       log15.Logger
+}
+
+func NewConnection(cfg *Config, kp *secp256k1.Keypair, log log15.Logger) *Connection {
+	signer := ethtypes.HomesteadSigner{}
+	return &Connection{
+		ctx: context.Background(),
+		cfg: *cfg,
+		// TODO: add network to use to config
+		signer:    signer,
+		kp:        kp,
+		nonceLock: sync.Mutex{},
+		log:       log,
+	}
+}
+
+// Connect starts the ethereum WS connection
+func (c *Connection) Connect() error {
+	c.log.Info("Connecting to ethereum chain...", "url", c.cfg.endpoint)
+	var rpcClient *rpc.Client
+	var err error
+	if c.cfg.http {
+		rpcClient, err = rpc.DialHTTP(c.cfg.endpoint)
+	} else {
+		rpcClient, err = rpc.DialWebsocket(c.ctx, c.cfg.endpoint, "/ws")
+	}
+	if err != nil {
+		return err
+	}
+
+	c.conn = ethclient.NewClient(rpcClient)
+	return nil
+}
+
+// Close stops the WS connection
+func (c *Connection) Close() {
+	c.conn.Close()
+}
+
+func (c *Connection) NetworkId() (*big.Int, error) {
+	return c.conn.NetworkID(c.ctx)
+}
+
+// subscribeToEvent registers an rpc subscription for the event with the signature sig for contract at address
+func (c *Connection) subscribeToEvent(query eth.FilterQuery) (*ActiveSubscription, error) {
+	ch := make(chan ethtypes.Log)
+	sub, err := c.conn.SubscribeFilterLogs(c.ctx, query, ch)
+	if err != nil {
+		close(ch)
+		return nil, err
+	}
+	return &ActiveSubscription{
+		ch:  ch,
+		sub: sub,
+	}, nil
+}
+
+// SubmitTx submits a transaction to the chain
+// It assumes the input data is an ethtypes.Transaction, marshalled as JSON
+func (c *Connection) SubmitTx(data []byte) error {
```

Comment:
I was thinking, could we capture the Nonce at this stage?
Furthermore, we might benefit from a queue, where we assign the Nonce as we fire them off.

Reply:
To be removed, we don't ever use `SubmitTx` #323